typescript/core/src/ChunkCursor.ts

```diff
@@ -211,80 +160,75 @@ export class ChunkCursor {
         continue;
       }
 
-      result.record.records.sort(([logTimeA], [logTimeB]) => Number(logTimeA - logTimeB));
-      if (reverse) {
-        // If we used `logTimeB - logTimeA` as the comparator for reverse iteration, messages with
-        // the same timestamp would not be in reverse order. To avoid this problem we use reverse()
-        // instead.
-        result.record.records.reverse();
-      }
-
-      for (const [logTime] of result.record.records) {
-        if (logTime < this.chunkIndex.messageStartTime) {
-          throw new Error(
-            `Encountered message index entry in channel ${result.record.channelId} with logTime (${logTime}) earlier than chunk messageStartTime (${this.chunkIndex.messageStartTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
-          );
-        }
-        if (logTime > this.chunkIndex.messageEndTime) {
-          throw new Error(
-            `Encountered message index entry in channel ${result.record.channelId} with logTime (${logTime}) later than chunk messageEndTime (${this.chunkIndex.messageEndTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
-          );
-        }
-      }
-
-      let startIndex = 0;
-      if (reverse) {
-        if (this.#endTime != undefined) {
-          startIndex = sortedIndexBy(result.record.records, this.#endTime, (logTime) => -logTime);
-        }
-      } else {
-        if (this.#startTime != undefined) {
-          startIndex = sortedIndexBy(result.record.records, this.#startTime, (logTime) => logTime);
-        }
-      }
-
-      if (startIndex >= result.record.records.length) {
-        continue;
-      }
-      if (reverse) {
-        if (
-          this.#startTime != undefined &&
-          result.record.records[startIndex]![0] < this.#startTime
-        ) {
-          continue;
-        }
-      } else {
-        if (this.#endTime != undefined && result.record.records[startIndex]![0] > this.#endTime) {
-          continue;
-        }
-      }
-
-      this.#messageIndexCursors.push({
-        index: startIndex,
-        channelId: result.record.channelId,
-        records: result.record.records,
-      });
+      this.#orderedMessageOffset = this.#orderedMessageOffset.concat(result.record.records);
     }
 
     if (offset !== messageIndexesView.byteLength) {
       throw new Error(
         `${messageIndexesView.byteLength - offset} bytes remaining in message index section`,
       );
     }
-  }
 
-  #getSortTime(): bigint {
-    if (!this.#messageIndexCursors) {
-      return this.#reverse ? this.chunkIndex.messageEndTime : this.chunkIndex.messageStartTime;
+    if (this.#orderedMessageOffset.length === 0) {
+      return;
+    }
+
+    this.#orderedMessageOffset.sort(([logTimeA], [logTimeB]) => Number(logTimeA - logTimeB));
+    if (reverse) {
```

Comment:
This is awkward - the message indexes in an MCAP are not specified as being in the same order as the messages, so you can't guarantee after sorting by timestamp that the messages with the same timestamp will be in file-offset order. This is also true in reverse.

I think you need to compare by log time and file offset when sorting.

Reply:
This seems like a good callout, but I don't think the behavior is changing here. It looks like the [heap we were previously using](https://github.com/foxglove/mcap/pull/1037/files#diff-cc829f0c49b6423567981a8ce9084d8d3d282c1faff58b0cd69830e8c43c7b59L143) didn't have a fallback comparator for file offset order.  On the other hand, we do have this in `compare()` already (for the higher-level heap across chunks): https://github.com/foxglove/mcap/blob/6986d8b32e9840fed5e7c18de7a20ce2fcd0a1c1/typescript/core/src/ChunkCursor.ts#L70